### PR TITLE
Fixed issue #1289

### PR DESCRIPTION
**Pull request by KalyanaChakravarthi**
Fixed issue of bad tokenization of double pointers

### CODE REVIEW

File: src/keywords.cpp [hunk 1]
```diff
@@ -88,6 +88,7 @@ static const chunk_tag_t keywords[] =
    { "__attribute__",      CT_ATTRIBUTE,     LANG_C | LANG_CPP                                                           },
    { "__block",            CT_QUALIFIER,     LANG_OC                                                                     },
    { "__const__",          CT_QUALIFIER,     LANG_C | LANG_CPP                                                           },
+   { "__declspec",         CT_DECLSPEC,      LANG_C | LANG_CPP | LANG_CS                                                 },
```
Comment (hschlichter) [inline src/keywords.cpp]: __declspec is a C/C++ specific modifier. Why is C# added as a language?

File: src/token_enum.h [hunk 2]
```diff
@@ -227,6 +227,7 @@ enum c_token_t
    CT_MACRO,            // a macro def
    CT_QUALIFIER,        // static, const, etc
    CT_EXTERN,           // extern
+   CT_DECLSPEC,         // __declspec
```
Comment (hschlichter) [inline src/token_enum.h:230]: It's way too specific to add as a token. If anything, it should be something about the storage class.

File: src/space.cpp [hunk 3]
```diff
@@ -168,6 +168,11 @@ static argval_t do_space(chunk_t *first, chunk_t *second, int &min_sp, bool comp
       return(AV_REMOVE);
    }
 
+   if (first->type == CT_DECLSPEC)
+   {
+      log_rule("IGNORE");
+      return(AV_IGNORE);
```
Comment (hschlichter) [inline src/space.cpp]: Why are you ignoring here? Doesn't make much sense.

**Review by hschlichter — CHANGES_REQUESTED** [review]
This doesn't at all solve the "bad tokenization of double pointers" issue.

As far as we can see, the reason why the travis and appveyor tests are green is because it ignores the function prototypes that have the __declspec. Which is super specific and wrong.

Furthermore, this causes major regressions in our code base, because the default behavior is changing in a dramatic way.

In a meeting some weeks ago, we talked about how to solve this. Currently uncrustify doesn't know enough about the type information when parsing function prototypes. This could be solved by running through the parsing a second time. We take this approach in several places. If you have any doubts about to actually fix this issue, reach out to us on #devs-uncrustify so we can talk about the solution.

**Comment by KalyanaChakravarthi** [comment]
test this please

**Comment by hschlichter** [comment]
There is still a diff on the unity-diff-checker. It removes a single line break. Which on the surface wouldn't seem like a big deal(and you can definitely argue that the line break should be removed). But the question is why does it do that?

**Comment by KalyanaChakravarthi** [comment]
Hi Henrik,
Can you please forward the test case for the above scenario?

**Comment by agandavarapu-rythmos** [comment]
@unitydevtools test

**Pull request merged**